test/pinverse.cpp

```diff
@@ -0,0 +1,327 @@
+    ASSERT_ARRAYS_NEAR(in, out, eps<TypeParam>());
+}
+
+TYPED_TEST(Pinverse, ApinvAApinv_Apinv) {
+    array in = readTestInput<TypeParam>(string(TEST_DIR"/pinverse/pinverse10x8.test"));
+    array inpinv = pinverse(in);
+    array out = matmul(inpinv, in, inpinv);
+    ASSERT_ARRAYS_NEAR(inpinv, out, eps<TypeParam>());
+}
+
+TYPED_TEST(Pinverse, AApinv_IsHermitian) {
+    array in = readTestInput<TypeParam>(string(TEST_DIR"/pinverse/pinverse10x8.test"));
+    array inpinv = pinverse(in);
+    array aapinv = matmul(in, inpinv);
+    array out = matmul(in, inpinv).H();
+    ASSERT_ARRAYS_NEAR(aapinv, out, eps<TypeParam>());
+}
+
+TYPED_TEST(Pinverse, ApinvA_IsHermitian) {
+    array in = readTestInput<TypeParam>(string(TEST_DIR"/pinverse/pinverse10x8.test"));
+    array inpinv = pinverse(in);
+    array apinva = af::matmul(inpinv, in);
+    array out = af::matmul(inpinv, in).H();
+    ASSERT_ARRAYS_NEAR(apinva, out, eps<TypeParam>());
+}
+
+TYPED_TEST(Pinverse, Large) {
+    array in = readTestInput<TypeParam>(string(TEST_DIR"/pinverse/pinverse640x480.test"));
+    array inpinv = pinverse(in);
+    array out = matmul(in, inpinv, in);
+    ASSERT_ARRAYS_NEAR(in, out, relEps<TypeParam>(in));
+}
+
+TEST(Pinverse, Square) {
+    array in = readTestInput<float>(string(TEST_DIR"/pinverse/pinverse10x10.test"));
+    array inpinv = pinverse(in);
+    array out = matmul(in, inpinv, in);
+    ASSERT_ARRAYS_NEAR(in, out, eps<float>());
+}
+
+TEST(Pinverse, Dim1GtDim0) {
+    array in = readTestInput<float>(string(TEST_DIR"/pinverse/pinverse8x10.test"));
+    array inpinv = pinverse(in);
+    array out = matmul(in, inpinv, in);
+    ASSERT_ARRAYS_NEAR(in, out, eps<float>());
+}
+
+TEST(Pinverse, CompareWithNumpy) {
+    array in = readTestInput<float>(string(TEST_DIR"/pinverse/pinverse10x8.test"));
+    array gold = readTestGold<float>(string(TEST_DIR"/pinverse/pinverse10x8.test"));
+    array out = pinverse(in);
+    ASSERT_ARRAYS_NEAR(gold, out, relEps<float>(gold));
+}
+
+TEST(Pinverse, SmallSigValExistsFloat) {
+    array in = readTestInput<float>(string(TEST_DIR"/pinverse/pinverse10x8.test"));
+    const dim_t dim0 = in.dims(0);
+    const dim_t dim1 = in.dims(1);
+
+    // Generate sigma with small non-zero value
+    af::array u;
+    af::array vT;
+    af::array sVec;
+    af::svd(u, sVec, vT, in);
+    dim_t sSize = sVec.elements();
+
+    sVec(2) = 1e-12;
+    af::array s = af::diag(sVec, 0, false);
+    af::array zeros = af::constant(0,
+                                   dim0 > sSize ? dim0 - sSize : sSize,
+                                   dim1 > sSize ? dim1 - sSize : sSize);
+    s = af::join(dim0 > dim1 ? 0 : 1, s, zeros);
+
+    // Make new input array that has a small non-zero value in its SVD sigma
+    in = af::matmul(u, s, vT);
+    array inpinv = pinverse(in);
+    array out = matmul(in, inpinv, in);
+
+    ASSERT_ARRAYS_NEAR(in, out, eps<float>());
+}
+
+TEST(Pinverse, SmallSigValExistsDouble) {
```

Comment:
This and the previous tests are almost identical. It would be nice if the logic was combined. Not important if it will take a long time.